engine/lib/users.php

```diff
@@ -28,25 +28,80 @@ function get_user(int $guid): ?\ElggUser {
 }
 
 /**
- * Get user by username
+ * Get a user by username
  *
- * @param string $username The user's username
+ * @param string $username  The user's username
+ * @param bool   $try_email If the username contains an '@' try based on email first
  *
- * @return \ElggUser|null Depending on success
+ * @return \ElggUser|null
+ * @since 5.0
  */
-function get_user_by_username(string $username): ?\ElggUser {
-	return _elgg_services()->usersTable->getByUsername($username);
+function elgg_get_user_by_username(string $username, bool $try_email = false): ?\ElggUser {
+	if (empty($username)) {
+		return null;
+	}
+	
+	if ($try_email && elgg_is_valid_email($username)) {
```

Comment:
maybe move this after the elgg_get_entities from username... the login action will allow try email, but mostly the login action is by username. This introduces an extra query... moving it after the username query will have this more like a fallback, as the primary login logic fails...

Reply:
Not going to change, the logic always was:
- if email try email
- else fallback to username

This way the logic stays intact